### ISSUE TITLE
Fails to compile with current liboqs version

### ISSUE THREAD

**Issue by geovandro** [open-quantum-safe/boringssl#64]
Tried to follow the standard instructions to compile BoringSSL with liboqs on Ubuntu and it seems liboqs has inconsistencies with BIKE. Would this be an issue?

The liboqs commit is 44cbde234a82813a181b4cc3c1330e56b253858f.

See the output below:
```bash
osboxes@osboxes:~$git clone --branch main --single-branch --depth 1 https://github.com/open-quantum-safe/liboqs.git
osboxes@osboxes:~$cd liboqs
osboxes@osboxes:~$mkdir build && cd build
osboxes@osboxes:~$cmake -G"Ninja" -DCMAKE_INSTALL_PREFIX=<MYBORINGSSL_DIR>/oqs -DOQS_USE_OPENSSL=OFF ..
osboxes@osboxes:~$ninja
[31/1941] Building C object src/kem/bike/CMakeFiles/bike_l3.dir/additional_r3/decode_avx512.c.o
FAILED: src/kem/bike/CMakeFiles/bike_l3.dir/additional_r3/decode_avx512.c.o 
/usr/bin/cc -DDISABLE_VPCLMUL -DFUNC_PREFIX=OQS_KEM_bike_l3 -DLEVEL=3 -DX86_64 -Iinclude -I../src/kem/bike/additional_r3 -fPIC -fvisibility=hidden   -march=native -Werror -Wall -Wextra -Wpedantic -Wstrict-prototypes -Wshadow -Wformat=2 -Wfloat-equal -Wwrite-strings -O3 -fomit-frame-pointer -fdata-sections -ffunction-sections -Wl,--gc-sections -mavx2 -include /home/osboxes/Documents/paranoid-pq/liboqs/src/kem/bike/functions_renaming.h -std=gnu11 -MD -MT src/kem/bike/CMakeFiles/bike_l3.dir/additional_r3/decode_avx512.c.o -MF src/kem/bike/CMakeFiles/bike_l3.dir/additional_r3/decode_avx512.c.o.d -o src/kem/bike/CMakeFiles/bike_l3.dir/additional_r3/decode_avx512.c.o   -c ../src/kem/bike/additional_r3/decode_avx512.c
../src/kem/bike/additional_r3/decode_avx512.c: In function ‘rotate512_big’:
../src/kem/bike/additional_r3/decode_avx512.c:43:21: error: AVX512F vector return without AVX512F enabled changes the ABI [-Werror=psabi]
       const __m512i a = LOAD(&out->qw[8 * (i + idx)]);
                     ^
In file included from /usr/lib/gcc/x86_64-linux-gnu/7/include/immintrin.h:45:0,
                 from ../src/kem/bike/additional_r3/x86_64_intrinsic.h:17,
                 from ../src/kem/bike/additional_r3/decode_avx512.c:25:
/usr/lib/gcc/x86_64-linux-gnu/7/include/avx512fintrin.h:461:1: error: inlining failed in call to always_inline ‘_mm512_mask_store_epi64’: target specific option mismatch
 _mm512_mask_store_epi64 (void *__P, __mmask8 __U, __m512i __A)
 ^~~~~~~~~~~~~~~~~~~~~~~
In file included from ../src/kem/bike/additional_r3/decode_avx512.c:25:0:
../src/kem/bike/additional_r3/x86_64_intrinsic.h:97:34: note: called from here
 #  define MSTORE(mem, mask, reg) _mm512_mask_store_epi64((mem), (mask), (reg))
                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/kem/bike/additional_r3/decode_avx512.c:44:7: note: in expansion of macro ‘MSTORE’
       MSTORE(&out->qw[8 * i], mask, a);
       ^~~~~~
In file included from /usr/lib/gcc/x86_64-linux-gnu/7/include/immintrin.h:45:0,
                 from ../src/kem/bike/additional_r3/x86_64_intrinsic.h:17,
                 from ../src/kem/bike/additional_r3/decode_avx512.c:25:
/usr/lib/gcc/x86_64-linux-gnu/7/include/avx512fintrin.h:5805:1: error: inlining failed in call to always_inline ‘_mm512_loadu_si512’: target specific option mismatch
 _mm512_loadu_si512 (void const *__P)
 ^~~~~~~~~~~~~~~~~~
../src/kem/bike/additional_r3/decode_avx512.c:43:21: note: called from here
       const __m512i a = LOAD(&out->qw[8 * (i + idx)]);
                     ^
In file included from /usr/lib/gcc/x86_64-linux-gnu/7/include/immintrin.h:45:0,
                 from ../src/kem/bike/additional_r3/x86_64_intrinsic.h:17,
                 from ../src/kem/bike/additional_r3/decode_avx512.c:25:
/usr/lib/gcc/x86_64-linux-gnu/7/include/avx512fintrin.h:461:1: error: inlining failed in call to always_inline ‘_mm512_mask_store_epi64’: target specific option mismatch
 _mm512_mask_store_epi64 (void *__P, __mmask8 __U, __m512i __A)
 ^~~~~~~~~~~~~~~~~~~~~~~
In file included from ../src/kem/bike/additional_r3/decode_avx512.c:25:0:
../src/kem/bike/additional_r3/x86_64_intrinsic.h:97:34: note: called from here
 #  define MSTORE(mem, mask, reg) _mm512_mask_store_epi64((mem), (mask), (reg))
                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/kem/bike/additional_r3/decode_avx512.c:44:7: note: in expansion of macro ‘MSTORE’
       MSTORE(&out->qw[8 * i], mask, a);
       ^~~~~~
In file included from /usr/lib/gcc/x86_64-linux-gnu/7/include/immintrin.h:45:0,
                 from ../src/kem/bike/additional_r3/x86_64_intrinsic.h:17,
                 from ../src/kem/bike/additional_r3/decode_avx512.c:25:
/usr/lib/gcc/x86_64-linux-gnu/7/include/avx512fintrin.h:5805:1: error: inlining failed in call to always_inline ‘_mm512_loadu_si512’: target specific option mismatch
 _mm512_loadu_si512 (void const *__P)
 ^~~~~~~~~~~~~~~~~~
../src/kem/bike/additional_r3/decode_avx512.c:43:21: note: called from here
       const __m512i a = LOAD(&out->qw[8 * (i + idx)]);
                     ^
cc1: all warnings being treated as errors
[33/1941] Building C object src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600times4_avx2.dir/KeccakP-1600times4/avx2/KeccakP-1600-times4-SIMD256.c.o
ninja: build stopped: subcommand failed.
```

**Comment by xvzcf** [comment]
Are you using a GCC version > 8?

**Comment by geovandro** [comment]
Yes, I recently moved to GCC 11.1 in order to reproduce and fix https://github.com/open-quantum-safe/liboqs/issues/1003

**Comment by geovandro** [comment]
Tried with GCC version 7.5.0 now, and had the same problems with BIKE.

**Comment by geovandro** [comment]
After some debugging I was able to figure it out: some files in BIKE implementation assume that every X86_64 architecture has AVX512 available, which is not the case. My i5 6267U processor for example has AVX2 but not AVX512. 

In order to make it compile in those platforms the [CMakeLists.txt](https://github.com/open-quantum-safe/liboqs/blob/main/src/kem/bike/CMakeLists.txt) file (lines 39-73) has to omit the declaration of source files *_avx512.c. In addition, files [sampling_internal.h](https://github.com/open-quantum-safe/liboqs/blob/main/src/kem/bike/additional_r3/sampling_internal.h), [decode_internal.h](https://github.com/open-quantum-safe/liboqs/blob/main/src/kem/bike/additional_r3/decode_internal.h) and [gf2x_internal.h](https://github.com/open-quantum-safe/liboqs/blob/main/src/kem/bike/additional_r3/gf2x_internal.h), have to omit the usage of the AVX512 functions.

After commenting those out, the code compiled.

**Comment by xvzcf** [comment]
> Yes, I recently moved to GCC 11.1 in order to reproduce and fix [open-quantum-safe/liboqs#1003](https://github.com/open-quantum-safe/liboqs/issues/1003)

Apologies, the `/usr/lib/gcc/x86_64-linux-gnu/7/` threw me off.
